consensus/src/experimental/buffer_manager.rs

```diff
@@ -204,33 +204,45 @@ impl StateManager {
         self.signing_root = cursor;
     }
 
-    /// check if a prefix of the buffer is ready to persist,
-    /// if yes, send them to the persisting phase and dequeue the buffer items.
-    fn try_persisting_blocks(&mut self) {
+    /// Pop the prefix of buffer items until (including) aggregated_item_cursor
+    /// Prepare an aggregated item and send it to the persisting phase
+    fn try_persisting_blocks(&mut self, aggregated_item_cursor: BufferItemRootType) {
         let mut cursor = self.buffer.head.as_ref().cloned();
+        let mut blocks_to_persist: Vec<Arc<ExecutedBlock>> = vec![];
+
         while cursor.is_some() {
             let buffer_item = take_elem(&cursor);
-            if let BufferItem::Aggregated(aggregated) = buffer_item {
-                let blocks: Vec<Arc<ExecutedBlock>> = aggregated
-                    .executed_blocks
-                    .into_iter()
-                    .map(Arc::new)
-                    .collect();
-                // send to persisting phase
-                block_on(self.persisting_phase_tx.send(PersistingRequest {
-                    blocks,
-                    commit_ledger_info: aggregated.aggregated_proof,
-                    callback: aggregated.callback,
-                }))
-                .ok();
-                cursor = get_next(&cursor);
-                self.buffer.pop_front();
-            } else {
-                // we put the item back
-                set_elem(&cursor, buffer_item);
+            blocks_to_persist.extend(
+                buffer_item
+                    .get_blocks()
+                    .iter()
+                    .map(|eb| Arc::new(eb.clone()))
+                    .collect::<Vec<Arc<ExecutedBlock>>>(),
+            );
+            if link_eq(&cursor, &aggregated_item_cursor) {
```

Comment:
I'd prefer we break out the loop first then send the request, seems more readable

Reply:
in this function I want to easily distinguish successfully finding the item vs reaching the end of the list, so I let it send the request here.

Reply:
hmm, it's not possible to reach the end of the list without hitting the cursor right?

Reply:
Right now it is not possible. Just in case of a bug in the future.

Reply:
it seems more natural to have something like
```
while curson.is_some() {
  if find {
    break 
  }
}
assert!(!blocks_to_persist.is_empty());
send_request
```

Reply:
`blocks_to_persist` collects the blocks before (including) the aggregated item. If it does not find the item, `blocks_to_persist` contains all the blocks. I added a flag `found` and moved the code outside the loop.

Reply:
oh, can't we just assert cursor.is_some() after the loop instead?

Reply:
We have to move out buffer_item anyway, so I just put it into an option and check if buffer_item has something. The code is already updated.

Reply:
I still feel we can simplify this code, but I guess I can try it myself after it gets merged